Documentation/articles/domain_driven_design/concepts_and_value_objects.md

```diff
@@ -0,0 +1,114 @@
+an address, etc..  It should be pointed out that the distinction between entity and value object can
+be domain specific.  What is a value object in one domain, might be an entity in another domain.
+
+
+# Primitive Obsession
+
+The term *primitive obsession* is used to describe the tendency to use language-specific  primitives
+(strings, integers, doubles, etc.) to represent concepts in the domain, rather than fully-fledged types.
+While ultimately all concepts must be expressed in the primitive types available in the language writing
+the domain-code like this leads to a less expressive, less correct domain model. For example, a telephone
+number might be represented as a *string* (or even an integer).  However, there is logic associated with
+a telephone number and aspects of it that are not necessarily captured by a string.  We may be interested
+in area codes, international dialling codes, extensions, country specific formatting rules, and so on.
+Replacing a primitive with a type allows us to capture that while not all strings are telephone numbers, 
+all telephone numbers can be expressed as strings.
+
+{{% notice warning %}}
+Never EVER represent things that look like "numbers" by virtue of being composed solely of digits as integers.
+{{% /notice %}} 
+
+# Value Objects in Dolittle
+
+Value object **MUST** be treated as immutable after creation and initial population.  Dolittle does not 
+enforce immutability, primarily as a concession to serialization and testing concerns.  However, it is 
+imperative that you do not alter any properties of a value object. Instead, you should always create and
+return a new instance.  Altering the properties of a Value Object will change the results of equality 
+comparisons and the result of *GetHashCode()*.
+
+Dolittle provides two different artifacts to support the creation of objects with value-identity semantics.
+
+# ConceptAs\<T>
+
+*ConceptAs\<T>* is intended as a direct replacement for where you would use a single primitive to represent
+a concept in the domain.  As well as implementing equality checks, and producing hash codes, *ConceptAs\<T>*
+allows an (almost) seamless transformation between the underlying primitive and the domain concept. This 
+proves to be very useful at the boundaries of your application, where you may be receiving primitives. It 
+can also aid in serialization and integration with other systems. *ConceptAs\<T>* implements an implicit 
+casting between the *ConceptAs\<T>* and the primitive *T*.
+
+A typical implementation of a ConceptAs might be
+
+```C#
+public class TelephoneNumber : ConceptAs<string>
+{
+    public static implicit operator TelephoneNumber(string value)
+    {
+        return new TelephoneNumber { Value = value };
+    }
+}
+```
+
+While the underlying ConceptAs<T> class can handle the implicit conversion to a string, it is not possible
+to implement the conversion to the specific type (i.e. TelephoneNumber) in a generic manner in the base class.
+
```

Comment:
Concept<T> now supports IComparable<Concept<T>> and IComparable and has also overloaded the operators < > <= >=.  This makes it much easier to replace the primitive with the concept.